### ISSUE TITLE
`throw Redirect()`

### ISSUE THREAD

**Issue by brillout** [brillout/vite-plugin-ssr#926]
### Description

It's arleady possible with https://vite-plugin-ssr.com/RenderErrorPage#redirection but the DX should be polished.

We should:
 - Create a second utility `throw Redirect()`
 - Make `pageContext.redirectTo` a built-in property
 - Update all server integration examples
 - Automatically redirect `/some/url/` to `/some/url`

**Comment by brillout** [comment]
Alternative design:

```js
import { Render } from 'vite-plugin-ssr/Render'

throw Render({
  redirect: '/login'
})
// Or render another page (while preserving the current URL), aka URL rewrite
throw Render({
  url: '/login'
})
// Or show the error page
throw Render({
  errorPage: true,
  // By default statusCode is 404 when `errorPage: true` and 307 when `redirect: '/some-url'`
  statusCode: 400
})
```

**Comment by magne4000** [comment]
I kinda like the [sveltekit](https://kit.svelte.dev/docs/load#redirects) [approach](https://kit.svelte.dev/docs/load#errors) on this matter:
```ts
import { error } from '@sveltejs/kit';

export function load() {
    throw error(401, 'not logged in');
}
```

```ts
import { redirect } from '@sveltejs/kit';

export function load() {
    throw redirect(307, '/login');
}
```

It enforces the status code, which is usually a good idea. No error are the same, and the most common error could not be a 404 on some app (same analogy goes for redirect).

I also like having specific functions for those different use case. It could just be some higher order functions that return a generic `Render` error.

tldr; I like this alternative design approach, I would just make it more explicit.

**Comment by brillout** [comment]
That makes a lot of sense. I'll think about it.

**Comment by brillout** [comment]
I think SvelteKit's design is impeccable on that one. Both succinct and clear. I'm 👍 for doing the same.

```js
/* We can't import from 'vite-plugin-ssr' in the 0.4.x release branch, we can move '/shared' to '/' in v1.0.0 
import { error } from 'vite-plugin-ssr'
*/
import { error } from 'vite-plugin-ssr/shared'

export function onSomeHook() {
    throw error(401, 'not logged in')
}
```

```js
import { redirect } from 'vite-plugin-ssr/shared'

export function onSomeHook() {
    throw redirect(307, '/login')
}
```
```js
import { rewrite } from 'vite-plugin-ssr/shared'

export function onSomeHook() {
    throw rewrite('/login')
}
```

I'll think about it some more and then start the implementation. Now that VPS has `guard()` hooks, this ticket is quite a must-have.

**Comment by brillout** [comment]
**Altneratives**

Slightly more explicit:

```js
import { redirect } from 'vite-plugin-ssr'

export function onSomeHook() {
    throw redirect(307, '/login')
}
```

```js
import { renderErrorPage } from 'vite-plugin-ssr'

export function onSomeHook() {
    throw renderErrorPage(401, 'not logged in')
}
```

```js
import { renderUrl } from 'vite-plugin-ssr'

export function onSomeHook() {
    throw renderUrl('/login')
}
```

Ugly but neat when using TypeScript:


```js
import { render } from 'vite-plugin-ssr'

export function onSomeHook() {
    throw render('redirect', 307, '/login')
}
```

```js
import { render } from 'vite-plugin-ssr'

export function onSomeHook() {
    throw render('error-page', 401, 'not logged in')
}
```

```js
import { render } from 'vite-plugin-ssr'

export function onSomeHook() {
    throw render('url', '/login')
}
```

**Comment by magne4000** [comment]
I like the "Slightly more explicit" version, not a fan of the "Ugly but neat when using TypeScript" one, mostly because it needs typescript annotations to be readable

**Comment by brillout** [comment]
List of relevant status codes: https://github.com/brillout/vite-plugin-ssr/issues/1008 (if the user doesn't use one of them then a warning is shown).

**Comment by brillout** [comment]
Quite a tempting alternative:

```js
import { redirect } from 'vite-plugin-ssr'

function onSomeHook() {
    throw redirect(307, '/login')
}
```

```js
import { render } from 'vite-plugin-ssr'

function onSomeHook() {
    throw render(401, 'not logged in')
}

function onSomeHook() {
    throw render('/login')
}
```

But let's see if it works out. (There are some slight argument conflicts.)

**Comment by brillout** [comment]
@magne4000 @gu-stav How about we remove the `statusCode` option from `throw redirect()`? The rationale being that if the redirection is done programmatically (dynamically) then it's most likely (always?) a temporary `302` redirect. At least I can think of a situation where I'd want to programmtically define a `301` permanent redirection.

So I think we can make `thow redirect()` always be a `302`, while adding a new feature for permanent redirections that the user statically defines:

```js
// /pages/+config.js

export default {
  // These are permanent 301 redirections
  redirect: [
    ['/about-us', '/about/team'],
    ['/vision', '/about/vision']
  ]
}
```

All in all, if I didn't miss any use case, it's both simpler and less error prone (choosing the wrong status code can have negative consequences that are quite substantial).

**Comment by gu-stav** [comment]
@brillout 👋🏼 

> How about we remove the statusCode option from throw redirect()?

I think there could be an argument for having `302` and `301`/`308` and therefore, I'd be in favor to keep them. Example: a very content-driven website wants to indicate whether something has been moved temporarily and permanently based on query params or the content itself.

I also think there is little overhead in this and making the arguments explicit even helps, because otherwise me as a developer would start to check the documentation to understand which status code is actually used.

**Comment by gu-stav** [comment]
> How about this then:

I think this us a good compromise! Keeps the API simple but allows full control for those who are in need 👍🏼

**Comment by brillout** [comment]
Done:
 - https://vite-plugin-ssr.com/redirect
 - https://vite-plugin-ssr.com/render

Leaving this open until permanent redirections with `config.redirects` are implemented.

**Comment by brillout** [comment]
> Leaving this open until permanent redirections with `config.redirects` are implemented.

Done: https://vite-plugin-ssr.com/redirects.

**Comment by brillout** [comment]
> how to disable the "Automatically redirect /some/url/ to /some/url" behavior? all urls must have a trailing slash in my app (external requirement)

https://github.com/brillout/vite-plugin-ssr/issues/949#issuecomment-1698985853
